doc/rzil.md

```diff
@@ -0,0 +1,132 @@
+partially relies on dynamic typing on C level, while still keeping static typing
+on IL level. In C, RzIL statically separates `RzILOpPure` and `RzILOpEffect`,
+but goes no further than that. The typedefs `RzILOpBitVector` and `RzILOpBool`
+are mere aliases of `RzILOpPure` to indicate which type is meant whenever it is
+known in the code. Further splitting is not possible while keeping ops like
+`ite` polymorphic.
+
+So far, this is only an implementation detail, but does not actually affect the
+language itself. However, as there are also certain limitations of OCaml's type
+system, our dynamic typing opens up other possibilities.
+
+For instance, the `append` op is used for appending two bitvectors of arbitrary
+sizes. Intuitively, the result's size would be the operands' numbers of bits
+added together, motivating a signature that would look somewhat like this:
+`append : 'b bitv -> 'c bitv -> ('b + 'c) bitv`. This however is simply not
+possible in OCaml, so BAP resorts to specify the result size explicitly as an
+argument and defines `append` to apply an extra cast after appending: `append :
+'a Bitv.t Value.sort -> 'b bitv -> 'c bitv -> 'a bitv` This solves the problem
+of typing, but makes the op's semantics somewhat more complicated.
+
+In contrast, RzIL's `append` always has the resulting size as the sum of the
+operand sizes, simplifying the semantics.
+
+### Variables
+
+Core Theory has three kinds of variables, which are categorized by the
+definition of their [identifier
+type](https://github.com/BinaryAnalysisPlatform/bap/blob/92d67c83fe0988b8a25bf563bdf33a9594db3e54/lib/bap_core_theory/bap_core_theory_var.ml#L20-L23):
+
+```ocaml
+type ident =
+  | Var of {num : Int63.t (* ... *)}
+  | Let of {num : Int63.t}
+  | Reg of {name : String.Caseless.t (* ... *)}
+```
+
+`Reg` vars, identified by a string, which usually correspond to physical
+registers of the emulated architecture, exist in the same way in RzIL as "global
+variables". They are defined as part of the vm setup and their scope is
+infinite.
+
+`Let` vars, which are immutable variables occuring as part of the pure `let` op,
+are implemented in a very similar way too. The only difference is that RzIL
+again uses string identifiers instead of integers. Their scope is naturally
+limited to the body of the `let` expression. As part of the RzIL vm, these
+variables are called "local pure variables".
+
+`Var` vars in Core Theory are so-called virtual variables, used mostly for
+temporary scratch locations. Their scope is unlimited, but when an instruction
+is lifted to Core Theory as part of the BAP Knowledge Base (this is
+approximately a very innovative kind of database),
+[fresh](https://github.com/BinaryAnalysisPlatform/bap/blob/92d67c83fe0988b8a25bf563bdf33a9594db3e54/lib/bap_core_theory/bap_core_theory_var.ml#L99-L101)
+temporary variables are always chosen with a globally unique identifier, thus
+automatically distinguishing temporary variables from multiple lifted
```

Comment:
Yes, I think there is a lot to learn from the KB approach. Still, I don't think this would apply as-is to variables like in BAP as there is a lot in KB that is pretty much impossible or impractical to have in C like the monadic code.